cmd/zed/zed.d/history_event-zfs-list-cacher.sh.in

```diff
@@ -0,0 +1,61 @@
+#!/bin/sh
+#
+# Track changes to specifically enumerated filesystems for system mounting.
+set -e
+
+FSLIST="@sysconfdir@/zfs/zfs-list.cache"
+FSLIST_TMP="@runstatedir@/zfs-list.cache.new"
+
+# If it is not writeable, then do not attempt to modify
+[ -w "${FSLIST}" ] || exit 0
+# If it is empty, abort
+[ "$(wc -l < "${FSLIST}")" -eq 0 ] && exit 0
+
+[ -f "${ZED_ZEDLET_DIR}/zed.rc" ] && . "${ZED_ZEDLET_DIR}/zed.rc"
+. "${ZED_ZEDLET_DIR}/zed-functions.sh"
+
+zed_exit_if_ignoring_this_event
+
+# We obtain a lock on zfs-list to avoid any simultaneous writes.
+# If we run into trouble, log and drop the lock
+abort_alter() {
+  zed_log_msg "Error updating zfs-list.cache!"
+  zed_unlock zfs-list
+}
+
+case "${ZEVENT_HISTORY_INTERNAL_NAME}" in
+  destroy|rename)
+   # because the history event only reports the command issued, rather than
+   # the affected ZFS, a rename or recursive destroy can affect a child ZFS.
+   # Instead of trying to figure out if we're affected, we instead just
+   # regenerate the cache *for that pool* on every modification.
+   ;;
+  set)
+   # only act if the mountpoint or canmount setting is altered
+    printf '%s' "${ZEVENT_HISTORY_INTERNAL_STR}" |
+      grep -q '^\(mountpoint\|canmount\)=' || exit 0
+   ;;
+  *)
+   # ignore all other events
+    exit 0
+   ;;
+esac
+
+zed_lock zfs-list
+trap abort_alter EXIT
+
+# Get the information for the affected pool
+grep "^${ZEVENT_POOL}[/$(printf '\t')]" "${FSLIST}" |
+grep -o '[^'"$(printf '\t')"'*' |
+  xargs @sbindir@/zfs list -H -oname,mountpoint,canmount \
+    >"${FSLIST_TMP}" 2>/dev/null || true
```

Comment:
depending on whether we want to react to any changes (see bigger comment above), the whole preceding section could be replaced with
`${ZFS} list -H -t filesystem -oname,mountpoint,canmount -r ${ZEVENT_POOL} > ${FSLIST_TMP}`